unity-renderer/Assets/DCLServices/WearablesCatalogService/LambdasWearablesCatalogService.cs

```diff
@@ -0,0 +1,308 @@
+        }
+
+        public void AddWearablesToCatalog(IEnumerable<WearableItem> wearableItems)
+        {
+            foreach (WearableItem wearableItem in wearableItems)
+            {
+                if (WearablesCatalog.ContainsKey(wearableItem.id))
+                    continue;
+
+                wearableItem.SanitizeHidesLists();
+                WearablesCatalog.Add(wearableItem.id, wearableItem);
+
+                if (!wearablesInUseCounters.ContainsKey(wearableItem.id))
+                    wearablesInUseCounters.Add(wearableItem.id, 1);
+            }
+        }
+
+        public void RemoveWearablesFromCatalog(IEnumerable<string> wearableIds)
+        {
+            foreach (string wearableId in wearableIds)
+            {
+                WearablesCatalog.Remove(wearableId);
+                wearablesInUseCounters.Remove(wearableId);
+            }
+        }
+
+        public void RemoveWearablesInUse(IEnumerable<string> wearablesInUseToRemove)
+        {
+            foreach (string wearableToRemove in wearablesInUseToRemove)
+            {
+                if (!wearablesInUseCounters.ContainsKey(wearableToRemove))
+                    continue;
+
+                wearablesInUseCounters[wearableToRemove]--;
+            }
+        }
+
+        public void EmbedWearables(IEnumerable<WearableItem> wearables)
+        {
+            foreach (WearableItem wearableItem in wearables)
+            {
+                WearablesCatalog[wearableItem.id] = wearableItem;
+
+                if (wearablesInUseCounters.ContainsKey(wearableItem.id))
+                    wearablesInUseCounters[wearableItem.id] = 10000; //A high value to ensure they are not removed
+            }
+        }
+
+        public void Clear()
+        {
+            WearablesCatalog.Clear();
+            wearablesInUseCounters.Clear();
+            awaitingWearableTasks.Clear();
+            pendingWearablesToRequest.Clear();
+        }
+
+        UniTask<(WearableResponse response, bool success)> ILambdaServiceConsumer<WearableResponse>.CreateRequest
+            (string endPoint, int pageSize, int pageNumber, CancellationToken cancellationToken) =>
+            lambdasService.Ref.Get<WearableResponse>(
+                "",
+                endPoint,
+                REQUESTS_TIME_OUT_SECONDS,
+                ATTEMPTS_NUMBER,
+                cancellationToken,
+                LambdaPaginatedResponseHelper.GetPageSizeParam(pageSize),
+                LambdaPaginatedResponseHelper.GetPageNumParam(pageNumber));
+
+        private async UniTask<WearableItem> SyncWearablesRequestsAsync(string newWearableId, CancellationToken ct)
+        {
+            if (!awaitingWearableTasks.TryGetValue(newWearableId, out var source))
+            {
+                pendingWearablesToRequest.Add(newWearableId);
+                awaitingWearableTasks[newWearableId] = source = new UniTaskCompletionSource<WearableItem>();
```

Comment:
One of the main reasons for my proposal was to avoid the creation of `UniTaskCompletionSource` per `WearableItem`. What you did here essentially means: We have a task that is related to `newWearableId` then we create one more `Task` inside the existing task to wait for the same `newWearableId`: it does not make much sense, does it?

Reply:
@mikhail-dcl I have apply the approach as you suggested initially but I need some help with a little detail. I understand why the first request is setting `lastRequestSource` to null BUT, because of this, it is making the rest of sent requests never ends. I mean, the current flow works in this way:
1. In the same frame we capture 3 requests.
https://github.com/decentraland/unity-renderer/blob/462934560cb767bd818cb7b5d631b93abbf09b33/unity-renderer/Assets/DCLServices/WearablesCatalogService/LambdasWearablesCatalogService.cs#L221

2. The first one creates a new `lastRequestSource`.
https://github.com/decentraland/unity-renderer/blob/462934560cb767bd818cb7b5d631b93abbf09b33/unity-renderer/Assets/DCLServices/WearablesCatalogService/LambdasWearablesCatalogService.cs#L225

3. The first one enters in the flow of `pendingWearablesToRequest.Count > 0` and then it caches the `lastRequestSource` into a local `source` and set `lastRequestSource` to null.
https://github.com/decentraland/unity-renderer/blob/462934560cb767bd818cb7b5d631b93abbf09b33/unity-renderer/Assets/DCLServices/WearablesCatalogService/LambdasWearablesCatalogService.cs#L228-L232

4. The first one send the request and keeps waiting.
https://github.com/decentraland/unity-renderer/blob/462934560cb767bd818cb7b5d631b93abbf09b33/unity-renderer/Assets/DCLServices/WearablesCatalogService/LambdasWearablesCatalogService.cs#L243-L248

5. The second and the third ones enter in the flow of `pendingWearablesToRequest.Count <= 0` and keep waiting for `lastRequestSource`.
https://github.com/decentraland/unity-renderer/blob/462934560cb767bd818cb7b5d631b93abbf09b33/unity-renderer/Assets/DCLServices/WearablesCatalogService/LambdasWearablesCatalogService.cs#L260-L261

6. The first request (who did the call to lambdas) finishes and return the result for that first wearable.
https://github.com/decentraland/unity-renderer/blob/462934560cb767bd818cb7b5d631b93abbf09b33/unity-renderer/Assets/DCLServices/WearablesCatalogService/LambdasWearablesCatalogService.cs#L265

8. But the other 2 wearables keep waiting indefinitely because `lastRequestSource` was set to null and will never resolves.

I have tried and if we don't set `lastRequestSource` to NULL, it works... but it would provoke a wrong behavior if we order more requests while the first one is still awaiting.

I can think of no other way to solve this than by managing a `UniTaskCompletionSource` by wearableId. If you come up with another solution, I'll be happy to try it.

Reply:
Done!